Data Preparation

In [1]:
import pandas as pd
import numpy as np

In [2]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [3]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

In [4]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [5]:
n_cols = predictors_norm.shape[1] # number of predictors

A. Build a baseline model
Use the Keras library to build a neural network with the following:
- One hidden layer of 10 nodes, and a ReLU activation function
- Use the adam optimizer and the mean squared error  as the loss function.

In [6]:
import keras

In [7]:

from keras.models import Sequential
from keras.layers import Dense

In [8]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))

    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [9]:
# build the model
model = regression_model()

1. Randomly split the data into a training and test sets by holding 30% of the data for testing.

In [10]:
import numpy as np
from sklearn.model_selection import train_test_split

In [11]:
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.30, random_state=42)

2. Train the model on the training data using 50 epochs.

In [12]:
# fit the model
model.fit(X_train, y_train, epochs=50, verbose=2)

Epoch 1/50
23/23 - 3s - loss: 1567.3716 - 3s/epoch - 152ms/step
Epoch 2/50
23/23 - 0s - loss: 1525.0948 - 160ms/epoch - 7ms/step
Epoch 3/50
23/23 - 0s - loss: 1470.2018 - 182ms/epoch - 8ms/step
Epoch 4/50
23/23 - 0s - loss: 1389.2866 - 90ms/epoch - 4ms/step
Epoch 5/50
23/23 - 0s - loss: 1272.4976 - 143ms/epoch - 6ms/step
Epoch 6/50
23/23 - 0s - loss: 1115.1168 - 233ms/epoch - 10ms/step
Epoch 7/50
23/23 - 0s - loss: 923.7480 - 118ms/epoch - 5ms/step
Epoch 8/50
23/23 - 0s - loss: 726.3514 - 98ms/epoch - 4ms/step
Epoch 9/50
23/23 - 0s - loss: 551.3511 - 79ms/epoch - 3ms/step
Epoch 10/50
23/23 - 0s - loss: 418.6889 - 90ms/epoch - 4ms/step
Epoch 11/50
23/23 - 0s - loss: 327.5624 - 150ms/epoch - 7ms/step
Epoch 12/50
23/23 - 0s - loss: 272.0669 - 143ms/epoch - 6ms/step
Epoch 13/50
23/23 - 0s - loss: 239.6760 - 154ms/epoch - 7ms/step
Epoch 14/50
23/23 - 0s - loss: 220.1241 - 100ms/epoch - 4ms/step
Epoch 15/50
23/23 - 0s - loss: 205.7115 - 104ms/epoch - 5ms/step
Epoch 16/50
23/23 - 0s - loss: 1

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength.

In [13]:
y_pred = model.predict(X_test)

10/10 [==============================] - 0s 4ms/step


In [14]:
from sklearn.metrics import mean_squared_error

In [15]:
mse = mean_squared_error(y_test, y_pred)

In [16]:
mse

127.07594586880211

4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

In [17]:
mse_list = []
for i in range(50):
  X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.30, random_state=42)
  model.fit(X_train, y_train, epochs=50)
  y_pred = model.predict(X_test)
  mse = mean_squared_error(y_test, y_pred)
  mse_list.append(mse)

Strumieniowane dane wyjściowe obcięte do 5000 ostatnich wierszy.
23/23 [==============================] - 0s 2ms/step - loss: 2612.7048
Epoch 27/50
23/23 [==============================] - 0s 2ms/step - loss: 2561.6895
Epoch 28/50
23/23 [==============================] - 0s 3ms/step - loss: 2508.8938
Epoch 29/50
23/23 [==============================] - 0s 4ms/step - loss: 2460.6445
Epoch 30/50
23/23 [==============================] - 0s 3ms/step - loss: 2414.6831
Epoch 31/50
23/23 [==============================] - 0s 3ms/step - loss: 2367.8872
Epoch 32/50
23/23 [==============================] - 0s 3ms/step - loss: 2327.5317
Epoch 33/50
23/23 [==============================] - 0s 3ms/step - loss: 2284.8054
Epoch 34/50
23/23 [==============================] - 0s 3ms/step - loss: 2239.1890
Epoch 35/50
23/23 [==============================] - 0s 3ms/step - loss: 2200.8074
Epoch 36/50
23/23 [==============================] - 0s 2ms/step - loss: 2161.7749
Epoch 37/50
23/23 [===============

5. Report the mean and the standard deviation of the mean squared errors.

In [18]:
mse_list

[1211.1393671615338,
 625.3952079341234,
 427.18637468242054,
 353.3272993687695,
 321.1584051571149,
 293.6566153234199,
 281.66600811882535,
 293.32671636744647,
 267.29230061292475,
 260.6928345920643,
 241.5468031780224,
 206.79636325848475,
 139.37424442973247,
 112.23742827095879,
 127.34476364756117,
 94.59419301824038,
 127.33461592542476,
 66.6379138940597,
 65.19532503991012,
 58.50364534093441,
 55.71850573211565,
 53.48465506342498,
 51.55306258410504,
 50.18404640249548,
 49.31325433457217,
 48.222665806204915,
 56.865313105870484,
 56.85755270095006,
 49.745794993434544,
 43.84860029621953,
 48.97981251842397,
 46.79763692104053,
 48.30821833110356,
 66.11983412197692,
 46.68005161739735,
 51.272641632432396,
 42.739651323431154,
 43.660460616409516,
 48.1329465697568,
 47.93579141747972,
 40.75528999045104,
 40.46775927542881,
 51.98952118577509,
 42.56656091705917,
 43.94391970348348,
 40.675285701894246,
 42.8384607443464,
 44.137702844229985,
 40.96661674833827,
 42.0

In [19]:
len(mse_list)

50

In [20]:
mean = sum(mse_list)/len(mse_list)

In [21]:
mean

140.22486695252778

In [22]:
std_dev_list = abs(mse_list - mean)
std_dev = sum(std_dev_list)/len(std_dev_list)

In [23]:
std_dev

124.01943569299266

How does the mean of the mean squared errors compare to that from Step B?

Part B:
mean mse from 50 models: 497.17

Part D: mean mse from 50 models: 140.22

The result in part B worse